### PR TITLE
Fix pad across processes dim in trainer and not being able to set the timeout

### LINKED ISSUE #24751: Stalled loop during prediction with deepspeed
### System Info

- `transformers` version: 4.31.0.dev0
- Platform: Linux-5.10.173-154.642.amzn2.x86_64-x86_64-with-glibc2.26
- Python version: 3.10.10
- Huggingface_hub version: 0.14.1
- Safetensors version: 0.3.1
- PyTorch version (GPU?): 2.0.0 (True)
- Tensorflow version (GPU?): not installed (NA)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using GPU in script?: Yes
- Using distributed or parallel set-up in script?: Yes

### Who can help?

@pacman100 (b/c deepspeed-only problem) @sgugger (b/c this is a documentation example script)

### Information

- [X] The official example scripts
- [ ] My own modified scripts

### Tasks

- [X] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

I am running `examples/pytorch/translation/run_translation.py` on a machine with 4 V100's.   To replicate my issue, run 

deepspeed examples/pytorch/translation/run_translation.py --deepspeed tests/deepspeed/ds_config_zero3.json /
--model_name_or_path t5-small /
--per_device_train_batch_size 1 /
--output_dir output_dir /
--overwrite_output_dir /
--fp16 /
--do_train /
--max_train_samples 64 /
--num_train_epochs 1 /
--dataset_name wmt16 /
--dataset_config "ro-en" /
--source_lang en /
--target_lang ro /
--do_predict  /
--max_predict_samples 64 /
--predict_with_generate

### Expected behavior

I would expect the script to fully run using `deepspeed`, not just without it.  Right now, it outputs an warning message 

```
Invalidate trace cache @ step 0: expected module 2, but got module 0
Invalidate trace cache @ step 1: expected module 116, but got module 2
```
and gets stuck during the `.evaluation_loop()` method.  I added some printing steps to the code, and it appeared that the code was stalling after the first `.prediction_step()`.

### PR DESCRIPTION
**Pull request by muellerzr**
# What does this PR do?

Reverts tiny regression where `dim=1` is needed during `pad_across_processes`, and `ddp_timeout` wasn't trickled down through `PartialState`

Fixes # (issue)

Solves https://github.com/huggingface/transformers/issues/24751

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@sgugger

### CODE REVIEW

**Review by sgugger — COMMENTED** [review]
I think it's `dim=1`, not `pad_index=1`. The `pad_index` is 0 by default, was that also the case before the integration with the Trainer?

**Review by sgugger — COMMENTED** [review]
And it seems like we use `-100` for the padding_index everywhere, so let's go with that.

**Review by sgugger — APPROVED** [review]

File: tests/trainer/test_trainer.py [hunk 1]
```diff
@@ -2098,6 +2099,51 @@ def test_no_wd_param_group(self):
...
+                "ro",
+                "--do_predict",
+                "--max_predict_samples",
+                "64",
+                "--predict_with_generate",
+                "--ddp_timeout",
+                "60",
+            ]
+            execute_subprocess_async(command)
+            # successful return here == success - any errors would have caused an error in the sub-call
```
Comment (sgugger) [inline tests/trainer/test_trainer.py:2145]:
```suggestion
            # successful return here == success - any errors would have caused an error or a timeout in the sub-call
```

**Review by sgugger — APPROVED** [review]
Thanks for adding a test!
